tests/turnstile_chain.c

```diff
@@ -0,0 +1,95 @@
+#include <ktest.h>
+#include <klog.h>
+#include <mutex.h>
+#include <runq.h>
+#include <sched.h>
+#include <thread.h>
+
+/* Length of the propagation chain. */
+#define T 4
+
+static mtx_t mtx[T + 1];
+/* For simpler code: propagator[0] = starter */
+static thread_t *propagator[T + 1];
+static thread_t *starter;
+
+static void set_prio(thread_t *td, prio_t prio) {
+  WITH_SPINLOCK(td->td_spin) {
+    sched_set_prio(td, prio);
+  }
+}
+
+// n <- [0..T]
+static int propagator_prio(int n) {
+  return n * RQ_PPQ;
+}
+
+/* - n <- [1..T]
+ * - mtx[n-1] is already owned by propagator[n-1] (it was run earlier)
+ * - propagator[n] acquires mtx[n] and blocks on mtx[n-1] causing
+ *   priority propagation to propagator[0..n-1]
+ */
+static void propagator_routine(int n) {
+  assert(thread_self()->td_prio == propagator_prio(n));
+  assert(mtx_owner(&mtx[n]) == NULL);
+  WITH_MTX_LOCK (&mtx[n]) {
+    assert(mtx_owner(&mtx[n - 1]) == propagator[n - 1]);
+    WITH_MTX_LOCK (&mtx[n - 1]) {
+      /* Nothing interesting here. */
+    }
+  }
+  assert(!td_is_borrowing(thread_self()));
+  assert(thread_self()->td_prio == propagator_prio(n));
+}
+
+static void starter_routine(void *_arg) {
+  assert(mtx_owner(&mtx[0]) == NULL);
+  WITH_MTX_LOCK (&mtx[0]) {
+    for (int i = 1; i <= T; i++) {
+      WITH_NO_PREEMPTION {
+        set_prio(propagator[i], propagator_prio(i));
+        sched_add(propagator[i]);
+        assert(thread_self()->td_prio == propagator_prio(i - 1));
+      }
+
+      /* propagator[i] waits for mtx[i-1] (owned by propagator[i-1]) */
+
+      /* Check if the priorities have propagated correctly. */
+      for (int j = 0; j < i; j++) {
+        assert(propagator[j]->td_prio == propagator_prio(i));
+        assert(td_is_borrowing(propagator[j]));
+      }
+    }
+  }
+  assert(thread_self()->td_prio == propagator_prio(0));
+  assert(!td_is_borrowing(thread_self()));
+}
+
+static int test_main(void) {
+  klog_setmask(KL_MASK(KL_TEST) | KL_MASK(KL_THREAD));
```

Comment:
Why do we need it here? And in line 90?

Reply:
Oh, that's to stop being flooded by random messages in klog. I'm not sure if it actually helps at all. I suppose I'll remove it since that's how it is in other tests.